src/Orleans/Messaging/Message.cs

```diff
@@ -601,7 +606,7 @@ private List<ArraySegment<byte>> Serialize_Impl(out int headerLengthOut, out int
             {
                 BufferPool.GlobalPool.Release(headerBytes);
             }
-            headerBytes = headerStream.ToBytes() as List<ArraySegment<byte>>;
+            headerBytes = context.Stream.ToBytes() as List<ArraySegment<byte>>;
             int headerLength = headerBytes.Sum(ab => ab.Count);
```

Comment:
This was not a change in this PR, but what about using context.Stream.CurrentOffset instead of the allocating enumerator in the Sum Linq extension method?

It is doable for both header and body here.